### DISCUSSION TITLE
System entities and instantiating prefabs

### DISCUSSION THREAD

**Discussion by galileolajara** [o3de/o3de#2273] [Help Wanted]
Hi! Does anyone know how to add components to system entities in O3DE? I can't seem to find any documentation about it. Assuming that the system entities is a lumberyard-thing, how can I make an entity to persist across changing of level (via "map" command)

I'm also having trouble looking for documentation on loading and spawning prefabs on runtime. I'm working on a game that's highly reliant on dynamic content (loading prefabs from the web), and I hope someone can share some advice on this.

Thanks!

**Comment by nick-l-o3de** [comment]
This deserves its own section in the docs, but its similar to Lumberyard.

In general, you extend the engine (add components) by putting their code in a gem
https://o3de.org/docs/user-guide/gems/development/overview/

That page glosses over the fact that the AZ::Module class has 2 different responsibilities
* Its constructor is supposed to initialize the list of all **component types** inside this gem module (ie, populate a list of their types so that the engine is aware of them and can register them with whatever they need to be registered with, call reflect, make them show up in the right menus, etc).   This includes singletons (System Components) as well as regular components of the kind which you put on entities in the world.  System Components and regular components are in fact all AZ::Components.
* Its GetRequiredSystemComponents virtual function is supposed to return a vector of all typeids of **system components** (ie, components to auto create on initialization and keep alive for the duration of the application).  As opposed to "Game" components which are the kind you attach to entities in the editor.  Any component typeids you return here will be added to a long-lived entity that is created and activated on app startup and deactivated on shutdown.  It will share the same entity with lots of other components from other gems, and the sort order of activation is defined by the dependencies your component declares in GetRequiredServices / GetProvidedServices 

Most gems come with at least 1 system component (the example game template for example does) but you can have as many as you want.  Recommend setting up a pure virtual API in a different include folder and talking to it via Interface<T> or EBUS.
